ragna/core/_rag.py

```diff
@@ -256,34 +256,19 @@ def _unpack_chat_params(
             for fn, model in component_models.items()
         }
 
-    async def _run(
+    def _run(
         self, fn: Union[Callable[..., T], Callable[..., Awaitable[T]]], *args: Any
-    ) -> T:
+    ) -> Awaitable[T]:
         kwargs = self._unpacked_params[fn]
-        if inspect.iscoroutinefunction(fn):
-            fn = cast(Callable[..., Awaitable[T]], fn)
-            coro = fn(*args, **kwargs)
-        else:
-            fn = cast(Callable[..., T], fn)
-            coro = run_in_threadpool(fn, *args, **kwargs)
-
-        return await coro
+        return as_awaitable(fn, *args, **kwargs)
```

Comment:
Previously, only `Chat` needed the ability to turn arbitrary functions into awaitables. Now, we also need this for the builtin assistants. Thus, this was factored out into `ragna._utils`. The same is true for (async) iterators below.